### PR TITLE
KAFKA-3615: Exclude test jars in kafka-run-class.sh

### PR DESCRIPTION
**Pull request by Ishiihara**
@granders @hachikuji Can you take a look when you have time? Appreciate your time to review.

### CODE REVIEW

File: bin/kafka-run-class.sh [hunk 1]
```diff
@@ -20,6 +20,15 @@ then
   exit 1
 fi
 
+check_file() {
+  file=$1
+  if [ -z "$(echo "$file" | grep "test")" ] ; then
```
Comment (ijuma) [inline bin/kafka-run-class.sh]: This seems a bit too broad, can we not make the check a bit stricter?

File: bin/kafka-run-class.sh [hunk 2]
```diff
@@ -20,6 +20,15 @@ then
   exit 1
 fi
 
+check_file() {
```
Comment (granthenke) [inline bin/kafka-run-class.sh]: I think the name could be more clear on whats being checked. Maybe `isTestJar` or something like that.

File: bin/kafka-run-class.sh (hunk 1 above)
Comment (granthenke) [inline bin/kafka-run-class.sh]: There are other files included too. Basically anything that gets built into `*/build/libs/*` will show up be matched. 

We should be sure to exclude javadoc, scaladoc, sources, test and test-sources jars.

**Comment by granthenke** [comment]
@Ishiihara Just want to confirm that this is only an issue when running from source right? When running from a release the only line that loads jars in `kafka-run-class.sh` is the following:

``` bash
# classpath addition for release
CLASSPATH=$CLASSPATH:$base_dir/libs/*
```

It shouldn't stop this change but I would love to have kafka-run-class.sh work the same for both source builds and release builds. If not done here I can open a jira to track that.

**Comment by Ishiihara** [comment]
@ijuma @granthenke Thanks for the review. I updated the patch to have more specific filtering condition and also make sure that we also filter files in case that we run from binary package.

**Comment by ewencp** [comment]
@Ishiihara In `build.gradle` we have a task called `systemTestLibs` that depends on some test jars. I think that might be because we have some tests that execute code that's in the test jars. I think @granders set that up so maybe he can clarify how that's used. My concern is that filtering out all the test jars might break some of those tests. If we need support for running some classes in test jars then we might need to make it possible to toggle this filtering on/off (though filtering out by default seems like a good choice).

**Comment by granders** [comment]
@Ishiihara @ewencp 
Yeah I have the same concern - this may break a few tests which depend on classes in the test jar - e.g. the stream smoke tests.

**Comment by ijuma** [comment]
`systemTestLibs` was originally added because we need `minikdc` for the system tests. And then streams also started using this (as @granders said). It seems to me that we should have a variable that needs to be set for the test jars to be included. It is unexpected to have that as the default behaviour in my opinion.

**Comment by granders** [comment]
An environmental variable to trigger addition of test jars seems reasonable

**Comment by ewencp** [comment]
@ijuma @granders Agreed on the default behavior.

**Comment by Ishiihara** [comment]
@ewencp @ijuma @granders Thanks for the review, I would add an environment variable to control test jar addition.

**Comment by Ishiihara** [comment]
Thanks for the review. Added a INCLUDE_TEST_JARS variable and change `filter_file` to `should_include_file` according to @granders suggestion.

**Comment by ewencp** [comment]
@Ishiihara but presumably this flag needs to be enabled in some of the system tests?

**Comment by granders** [comment]
@Ishiihara Offhand, I think Minikdc service and streams-related services will need this flag (in system test code)

And then it would probably be good to run in a branch builder

**Comment by Ishiihara** [comment]
@granders Added include test jars to relevant system test services.

File: bin/kafka-run-class.sh [hunk 3]
```diff
@@ -80,22 +108,33 @@ for cc_pkg in "api" "runtime" "file" "json" "tools"
...
 for file in $base_dir/core/build/libs/kafka_${SCALA_BINARY_VERSION}*.jar;
 do
-  CLASSPATH=$CLASSPATH:$file
+  if should_include_file "$file"; then
+    CLASSPATH=$CLASSPATH:$file
+  fi
 done
 shopt -u nullglob
 
+echo $CLASSPATH
```
Comment (ewencp) [inline bin/kafka-run-class.sh]: Leftover debug output

**Comment by ewencp** [comment]
This LGTM, I'll just remove that one extra `echo` during commit

**Comment by Ishiihara** [comment]
@ewencp Thanks for the review, removed the left over debug output.

File: bin/kafka-run-class.sh [hunk 4]
```diff
@@ -80,19 +108,28 @@ for cc_pkg in "api" "runtime" "file" "json" "tools"
...
+    fi
   done
   if [ -d "$base_dir/connect/${cc_pkg}/build/dependant-libs" ] ; then
     CLASSPATH=$CLASSPATH:$base_dir/connect/${cc_pkg}/build/dependant-libs/*
   fi
 done
 
 # classpath addition for release
-CLASSPATH=$CLASSPATH:$base_dir/libs/*
+for file in $base_dir/libs;
```
Comment (dpkp) [inline bin/kafka-run-class.sh:121]: this is broken. It should be `for file in $base_dir/libs/*;` ... without the glob, you're looping over 1 entry: "libs".

I believe this bug breaks the classpath for all scripts in bin/ in the release artifact

**Pull request closed without merging**